src/hotspot/cpu/aarch64/assembler_aarch64.hpp

```diff
@@ -2410,6 +2468,57 @@ void mvnw(Register Rd, Register Rm,
 #undef INSN3
 #undef INSN4
 
+// Define a set of INSN1/2/3/4 macros to handle single-structure
+// load/store instructions.
+#define INSN1(NAME, op1, op2)                                           \
+  void NAME(FloatRegister Vt, SIMD_RegVariant T, int index,             \
+            const Address &a) {                                         \
+    assert(index >= 0 && (T <= D) && ((T == B && index <= 15) ||        \
+                (T == H && index <= 7) || (T == S && index <= 3) ||     \
+                (T == D && index <= 1)), "invalid index");              \
+    ld_st(Vt, T, index, a, op1, op2, 1);                                \
+ }
+
+#define INSN2(NAME, op1, op2)                                           \
+  void NAME(FloatRegister Vt, FloatRegister Vt2, SIMD_RegVariant T,     \
+            int index, const Address &a) {                              \
+    assert(index >= 0 && (T <= D) && ((T == B && index <= 15) ||        \
+                (T == H && index <= 7) || (T == S && index <= 3) ||     \
+                (T == D && index <= 1)), "invalid index");              \
+    assert(Vt->successor() == Vt2, "Registers must be ordered");        \
+    ld_st(Vt, T, index, a, op1, op2, 2);                                \
+  }
+
+#define INSN3(NAME, op1, op2)                                           \
+  void NAME(FloatRegister Vt, FloatRegister Vt2, FloatRegister Vt3,     \
+            SIMD_RegVariant T, int index, const Address &a) {           \
+    assert(index >= 0 && (T <= D) && ((T == B && index <= 15) ||        \
+                (T == H && index <= 7) || (T == S && index <= 3) ||     \
+                (T == D && index <= 1)), "invalid index");              \
+    assert(Vt->successor() == Vt2 && Vt2->successor() == Vt3,           \
+           "Registers must be ordered");                                \
+    ld_st(Vt, T, index, a, op1, op2, 3);                                \
+  }
+
+#define INSN4(NAME, op1, op2)                                           \
+  void NAME(FloatRegister Vt, FloatRegister Vt2, FloatRegister Vt3,     \
+            FloatRegister Vt4, SIMD_RegVariant T, int index,            \
+            const Address &a) {                                         \
+    assert(index >= 0 && (T <= D) && ((T == B && index <= 15) ||        \
+                (T == H && index <= 7) || (T == S && index <= 3) ||     \
+                (T == D && index <= 1)), "invalid index");              \
+    assert(Vt->successor() == Vt2 && Vt2->successor() == Vt3 &&         \
+           Vt3->successor() == Vt4, "Registers must be ordered");       \
+    ld_st(Vt, T, index, a, op1, op2, 4);                                \
+  }
+
+  INSN4(st4, 0b001101001, 0b0010);
+
+#undef INSN1
+#undef INSN2
+#undef INSN3
+#undef INSN4
+
```

Comment:
This code to handle the AdvSIMD load/store single structure and AdvSIMD load/store single structure (post-indexed) is excessive.

Every one of these instructions has the the format, 

`0|Q|0011010|L|R|00000|opcode|S|size|Rn|Rt`

or

`0|Q|0011011|L|R|   Rm|opcode|S|size|Rn|Rt`

Perhaps consider using a `RegSet regs` for the registers. Then the instruction encoding to use (1,2,3,or 4 consecutive registers) can be picked up from `regs.size()`. There only needs to be a single routine for all of the `ld_st` variants.

Reply:
Thanks for the suggestion.  I will look into this.  I can see how `regs.size()` could simplify these macros.

Reply:
Another thing that may be better than a `RegSet`. If you use a C++11 template parameter pack, you can do something like this:

```
template<typename R, typename... Rx>
void foo(R first_register, Rx... more_registers) {
  const R regs[] = { first_register, more_registers... };  // An array that contains the more regs
  const int count = sizeof...(more_registers);             // The count of more regs
  ...
}
```
And then you can use the same logic, regardless of the number of registers.